Mod/Classes/Patched/MyPlayerGhost.cs

```diff
@@ -3,58 +3,235 @@
 using TowerFall;
 using Monocle;
 using System;
+using New;
 
 namespace Mod
 {
   [Patch]
   public class MyPlayerGhost : PlayerGhost
   {
-    PlayerCorpse corpse;
-
-    public MyPlayerGhost(PlayerCorpse corpse)
-      : base(corpse)
-    {
-      this.corpse = corpse;
-    }
-
-    public override void Die(int killerIndex, Arrow arrow, Explosion explosion, ShockCircle circle)
-    {
-      base.Die(killerIndex, arrow, explosion, circle);
-      var mobLogic = this.Level.Session.RoundLogic as MobRoundLogic;
-      if (mobLogic != null) {
-        // Ghosts treated as players in crawl mode
-        mobLogic.OnPlayerDeath(
-          null, this.corpse, this.PlayerIndex, DeathCause.Arrow, // FIXME
-          this.Position, killerIndex
-        );
-      }
-    }
-
-    public override void Hurt (Vector2 force, int damage, int killerIndex, Arrow arrow = null, Explosion explosion = null, ShockCircle shock = null)
-    {
-      if (shock && killerIndex == this.PlayerIndex) {
-        // ShockCircle shouldn't kill friendly ghosts
-        return;
-      }
-
-      this.Speed = force;
-      if (this.Alive && this.CanHurt) {
-        this.Health -= damage;
-        if (this.Health <= 0) {
-          this.Die (killerIndex, arrow, explosion, shock);
-        }
-      }
-    }
-
-    public override void Update()
-    {
-      if (((MyMatchVariants)Level.Session.MatchSettings.Variants).FastGhosts) {
-        typeof(Engine).GetProperty("TimeMult").SetValue(null, Engine.TimeMult * 1.5f, null);
-        base.Update();
-        typeof(Engine).GetProperty("TimeMult").SetValue(null, Engine.TimeMult / 1.5f, null);
-      } else {
-        base.Update();
-      }
-    }
-  }
+        PlayerCorpse corpse;
+		public bool HasSpeedBoots;
+		public bool Invisible;
+		public PlayerGhostShield shield;
+		public WrapHitbox shieldHitbox;
+		public Counter shieldRegenCounter;
+		public float InvisOpacity = 1f;
+
+		public Scheduler scheduler;
+
+        public MyPlayerGhost(PlayerCorpse corpse)
+          : base(corpse)
+        {
+			this.corpse = corpse;
+            this.Allegiance = corpse.Allegiance;
+			this.shield = new PlayerGhostShield(this);
+			this.shieldHitbox = new WrapHitbox(16f, 18f, -8f, -10f);
+			base.Add(this.shield);
+        }
+
+    	public MyPlayerGhost(PlayerCorpse corpse, PlayerGhostInventory inventory)
+            : base(corpse)
+		{
+            this.corpse = corpse;
+
+			this.shieldHitbox = new WrapHitbox(16f, 18f, -8f, -10f);
+            this.shieldRegenCounter = new Counter(240);
+			this.shield = new PlayerGhostShield(this);
+
+    		if (inventory.Shield)
+            {
+                this.scheduler.ScheduleAction(delegate {
+                    this.HasShield = true;
+                }, 20 + 10 * this.PlayerIndex, false);
```

Comment:
Multiplying by PlayerIndex seems weird here. Did you copy that from somewhere? If not, what's it doing?

Reply:
That was part of the inventory that wasn't actually needed for this feature. It came direct from the source code though. I have no idea what it does.